inception/inception-api-annotation-experimental/src/main/java/de/tudarmstadt/ukp/inception/experimental/api/AnnotationSystemAPIImpl.java

```diff
@@ -17,402 +17,669 @@
+    {
+        try {
+            List<AnnotationLayer> layers = annotationService
+                    .listAnnotationLayer(projectService.getProject(aProjectId));
 
-                for (Character c : textInSentences[aClientMessage.getViewport()[i][0]]
-                        .toCharArray()) {
-                    visibleSentences.add(c);
+            for (AnnotationLayer layer : layers) {
+                if (aLayer.equals(layer.getUiName()) || aLayer.equals(layer.getName())) {
+                    return layer;
                 }
-
-                visibleSentences.add('|');
             }
-            return visibleSentences.toArray(new Character[0]);
-        default:
-            System.err.println("Offset type not found");
         }
+        catch (Exception e) {
+            LOG.error("Exception has been thrown during getLayer()", e);
+            return null;
+        }
+
+        LOG.error("Exception has been thrown during getLayer(): Not a layer with the name " + aLayer
+                + " found.");
         return null;
     }
 
-    @Override
-    public List<Annotation> getAnnotations(CAS aCas, long aProject)
+    public List<AnnotationLayer> getValidLayers(long aProjectId, List<String> disabledLayers)
     {
-        List<AnnotationListItem> items = new ArrayList<>();
-        List<AnnotationLayer> metadataLayers = annotationService
-            .listAnnotationLayer(projectService.getProject(aProject));
+        if (disabledLayers.isEmpty()) {
+            return annotationService.listAnnotationLayer(projectService.getProject(aProjectId));
+        }
 
-        for (AnnotationLayer layer : metadataLayers) {
-            if (layer.getUiName().equals("Token")) {
-                // TODO: exception later when calling renderer.getFeatures "lemma"
-                continue;
+        List<AnnotationLayer> validLayers = new ArrayList<>();
+        List<AnnotationLayer> layers = annotationService
+                .listAnnotationLayer(projectService.getProject(aProjectId));
+        for (AnnotationLayer layer : layers) {
+            if (!disabledLayers.contains(layer.getType())) {
+                validLayers.add(layer);
             }
+        }
+        return validLayers;
+    }
 
-            TypeAdapter adapter = annotationService.getAdapter(layer);
-
-            for (FeatureStructure fs : selectFS(aCas, adapter.getAnnotationType(aCas))) {
-                String labelText = "";
-                items.add(new AnnotationListItem(WebAnnoCasUtil.getAddr(fs), labelText, layer));
+    public List<Span> filterSpans(List<Span> aAnnotations, Viewport aViewport) throws Exception
+    {
+        try {
+            List<Span> filteredAnnotations = new ArrayList<>();
+            for (Span annotation : aAnnotations) {
+                for (int i = 0; i < aViewport.getViewport().size(); i++) {
+                    for (int j = aViewport.getViewport().get(i).get(0); j <= aViewport.getViewport()
```

Comment:
If I see this correctly, you it iterate over every position in the viewport and then check if any span annotation starts at that position - if that is true, it would be really inefficient.

Also, how do you handle spans that partially overlap or fully include the viewport? 

Have a look at `de.tudarmstadt.ukp.clarin.webanno.api.annotation.rendering.SpanRenderer.render(CAS, List<AnnotationFeature>, VDocument, int, int)`